File: hard/Lc135.py
"""
2022年4月7日
1 递增递减序列
2 O(N)
3 30min
"""
class Lc164:
    def candy(self, ratings: List[int]) -> int:
        # 前后比较的问题一般i都从1开始
        i, ans, ls, last = 1, 0, len(ratings), 0
        candys = [1 for i in range(lr)]
        while i < ls:
            # 递增序列
            if ratings[i] > ratings[i-1]:
                candys[i] = candys[i-1]+1
                i += 1
            # 不增不减序列
            elif ratings[i] == ratings[i-1]:
                candys[i] = 1
                i += 1
            else:
                # 递减序列
                last = i-1
                while i < ls and ratings[i] < ratings[i-1]:
                    i += 1
                # 注意此时i-1才是递减序列的末尾，递减序列的范围[last, i-1]
                # 特别的，i=ls不影响后续的逻辑
                candys[last] = max(candys[last], i-last)
                candys[i-1] = 1
                for j in range(i-2, last, -1):
                    candys[j] = candys[j+1] + 1
        for c in candys:
            ans += c
        return ans

File: hard/Lc135/Lc135.go
package Lc135

// 极值问题 a b c
// 贪心
// dp
// 递增递减序列
// lastPeek length i j up down
// 1 2 3 3 2 1 2
//       j     i
//           1
func candy(ratings []int) int {
	if len(ratings) == 1 {
		return 1
	}
	i, ans, last := 1, 1, 1
	for i < len(ratings) {
		if ratings[i-1] < ratings[i] {
			last += 1
			ans += last
			i += 1
		} else if ratings[i-1] == ratings[i] {
			last = 1
			ans += last
			i += 1
		} else {
			j := i - 1
			// 	ratings[i-1] > ratings[i]
			for i < len(ratings) && ratings[i-1] > ratings[i] {
				i += 1
			}
			// [j, i-1]
			k := 1
			for ; k < i-j; k++ {
				ans += k
			}
			if k > last {
				ans += k - last
			}
			last = 1
		}
	}
	return ans
}
